data_diff/diff_tables.py

```diff
@@ -183,52 +182,47 @@ def diff_tables(self, table1: TableSegment, table2: TableSegment) -> DiffResult:
         if self.bisection_factor < 2:
             raise ValueError("Must have at least two segments per iteration (i.e. bisection_factor >= 2)")
 
-        return self._diff_tables(table1, table2)
+        logger.info(
+            f"Diffing tables | segments: {self.bisection_factor}, bisection threshold: {self.bisection_threshold}."
+        )
 
-    def _diff_tables(self, table1, table2, level=0, segment_index=None, segment_count=None):
-        count1, count2 = self._precalc_attr_threaded("count", [table1, table2])
+        key_ranges = self._threaded_call("query_key_range", [table1, table2])
+        mins, maxs = zip(*key_ranges)
+
+        # We add 1 because our ranges are exclusive of the end (like in Python)
+        start_key = min(mins)
+        end_key = max(maxs) + 1
+
+        table1 = table1.new(start_key=start_key, end_key=end_key)
+        table2 = table2.new(start_key=start_key, end_key=end_key)
+
+        return self._bisect_and_diff_tables(table1, table2)
 
-        if segment_index:
-            logger.info(". " * level + f"Diffing segment {segment_index}/{segment_count} of size {count1} and {count2}")
-        else:
-            logger.info(
-                f"Diffing tables of size {table1.count} and {table2.count} | segments: {self.bisection_factor}, bisection threshold: {self.bisection_threshold}."
-            )
+    def _bisect_and_diff_tables(self, table1, table2, level=0, max_rows=None):
+        assert table1.is_bounded and table2.is_bounded
 
-        checksum1, checksum2 = self._precalc_attr_threaded(
-            "checksum", [table1, table2]
-        )  # Calculate checksum in parallel
+        if max_rows is None:
```

Comment:
Ah, I didn't see that you passed `max(count1, count2)` here when recursing. Took me a second. I thought for a bit that you'd done away with using `count` here beyond at level > 0.

I think the recursion would be clearer if you just passed `table1.end_key - table1.start_key` from the caller for the first iteration 👍🏻